pkg/daemon/daemon.go

```diff
@@ -177,38 +226,116 @@ func (dn *Daemon) CheckStateOnBoot(stop <-chan struct{}) error {
 	return nil
 }
 
-// handleNodeUpdate is the handler for informer callbacks detecting node
-// changes. If an update is requested by the controller, we assume that
-// that means something changed and apply the desired config no matter what.
+// runOnce pulls the MachineConfig from either a file or remote URL
+// executes one run, and exits.
+// TODO: Revisit the Run/process methods and refactor and unify them with runOnce
+func (dn *Daemon) runOnce() error {
+	var machineConfig *mcfgv1.MachineConfig
+	var oldConfig mcfgv1.MachineConfig
+	var err error
+
+	if strings.HasPrefix("http://", dn.onceFrom) || strings.HasPrefix("https://", dn.onceFrom) {
+		// NOTE: This case expects a cluster to exists already.
+		// If we sense a remote URL has been provided then request MC content
+		// from a remote URL and parse it.
+		glog.V(2).Infof("Getting machine config content from %s", dn.onceFrom)
+		machineConfig, err = dn.getMachineConfigFromURL(dn.onceFrom)
+		if err != nil {
+			return err
+		}
+		needUpdate, err := dn.prepUpdateFromCluster()
+		if err != nil {
+			glog.V(2).Infof("Unable to prep update: %s", err)
+			return err
+		} else if needUpdate == false {
+			glog.V(2).Infof("No update needed")
+			return nil
+		}
+		// At this point we have verified we need to update
+		if err := dn.executeUpdateFromClusterWithMachineConfig(machineConfig); err != nil {
+			glog.Warningf("Unable to update: %s", err)
+			return err
+		}
+		return nil
+
+	} else if ValidPath(dn.onceFrom) {
+		// NOTE: This case expects that the cluster is NOT CREATED YET.
+		// If we sense a local file has been provided parse it.
+		oldConfig = mcfgv1.MachineConfig{}
+		absoluteOnceFrom, err := filepath.Abs(filepath.Clean(dn.onceFrom))
+		if err != nil {
+			return err
+		}
+		machineConfig, err = dn.getMachineConfigFromFile(absoluteOnceFrom)
+		if err != nil {
+			return err
+		}
+		// Execute update without hitting the cluster
+		return dn.update(&oldConfig, machineConfig)
+	}
+	// Otherwise return an error as the input format is unsupported
+	return fmt.Errorf("%s is not a path nor url; can not run once", dn.onceFrom)
+}
+
+// handleNodeUpdate is the gatekeeper handler for informer callbacks detecting
+// node changes. If an update is requested by the controller, we assume that
+// that means something changed and pass over to execution methods no matter what.
 // Also note that we only care about node updates, not creation or deletion.
 func (dn *Daemon) handleNodeUpdate(old, cur interface{}) {
 	node := cur.(*corev1.Node)
 
 	// First check if the node that was updated is this daemon's node
-	if (node.Name != dn.name) {
-		// The node that was changed was not ours
-		return
+	if node.Name == dn.name {
+		// Pass to the shared update prep method
+		_, err := dn.prepUpdateFromCluster()
+		if err != nil {
+			glog.V(2).Infof("Unable to prep update: %s", err)
+			return
+		}
+		if err = dn.executeUpdateFromCluster(); err != nil {
```

Comment:
If there is no update from the above `prepUpdateFromCluster`, shouldn't we not run this? Currently we only return if that previous function errors.

Also, we should be passing the `desiredConfig` from the `prepUpdate` above, otherwise we'll hit https://github.com/openshift/machine-config-operator/pull/145 again, right?

Reply:
Agreed on first point.

On the second point are you recommending that we should return the `desiredConfig` as a return value for use later?

Reply:
Actually maybe I mis-understood the update method the first time around, I think this should be fine.

Originally, the main concern being that let's say 2 updates happen in quick succession, then the first update is not stored, and we might see the second `desiredConfig` when we poll it again at: https://github.com/openshift/machine-config-operator/pull/126/files#diff-06b532a2a5a5d3ad1d5341b3e00945efR404 , but that should be fine since we'd just apply 2 updates and bring it to the most updated "desired state"? Correct me if I'm wrong in this understanding. If that's how it operates, then it's fine to leave as is.

Reply:
`prepUpdateFromCluster` is:

Reply:
Yes, I mistook the application of config and updating the annotation. This should be fine.